src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -0,0 +1,646 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::{Ordering, Reverse};
+use std::collections::BinaryHeap;
+use std::fmt::Debug;
+use std::ops::Range;
+
+use datatypes::arrow::array::{
+    ArrayRef, TimestampMicrosecondArray, TimestampMillisecondArray, TimestampNanosecondArray,
+    TimestampSecondArray, UInt64Array,
+};
+use datatypes::arrow::datatypes::{DataType, TimeUnit};
+
+use crate::error::Result;
+use crate::memtable::merge_tree::data::{DataBatch, DataBufferReader, DataPartReader};
+use crate::memtable::merge_tree::PkIndex;
+
+/// Nodes of merger's heap.
+pub trait Node: Ord {
+    type Item;
+
+    /// Returns current item of node and fetch next.
+    fn fetch_next(&mut self) -> Result<Self::Item>;
+
+    /// Returns true if current node is not exhausted.
+    fn is_valid(&self) -> bool;
+
+    /// Current item of node.
+    fn current_item(&self) -> &Self::Item;
+
+    /// Whether the other node is behind (exclusive) current node.
+    fn is_behind(&self, other: &Self) -> bool;
+
+    /// Skips first `num_to_skip` rows from node's current batch. If current batch is empty it fetches
+    /// next batch from the node.
+    ///
+    /// # Panics
+    /// If the node is EOF.
+    fn skip(&mut self, offset_to_skip: usize) -> Result<()>;
+
+    /// Searches given item in node's current item and returns the index.
+    fn search_key_in_current_item(&self, key: &Self::Item) -> std::result::Result<usize, usize>;
+
+    /// Slice current item.
+    fn slice_current_item(&self, range: Range<usize>) -> Self::Item;
+}
+
+pub struct Merger<T, I> {
+    heap: BinaryHeap<T>,
+    current_item: Option<I>,
+}
```

Comment:
We can reduce the number of generic parameters
```suggestion
pub struct Merger<T: Node> {
    heap: BinaryHeap<T>,
    current_item: Option<T::Item>,
}
```